### ISSUE TITLE
postgres & sequelize support

### ISSUE THREAD

**Issue by musicformellons** [stalniy/casl#8]
I am a bit puzzled by the following:
1) As my app has returning users with account etc I suppose I do need a database for authorization rules etc.
2) I really would prefer to use postgres (with sequelize) instead of mongodb. I saw some code examples in the docs for postgres & sequelize but it feels a bit like this is not really supported (yet) and I find it hard to judge whether I could get it up and running with these pointers. I guess my question boils down to: are these given examples enough to be up and running with postgres and sequelize, or are they just boilerplate   examples on which more work is needed before you can use it?

**Comment by stalniy** [comment]
Integration with database should be quite easy because the only thing you need to do is to convert CASL rules into database query. `sequelize` uses similar syntax to construct SQL query. So, in terms of fetching all records according to ACL rules from the single table shouldn't be a big deal and examples from documentation should cover majority of such cases.

So, when you define permissions for an entity based on its own fields, all should be good. 
But CASL doesn't provide a solution if you need to define permissions for one entity based on fields of another entity and it doesn't matter which database you use. So, for such cases you will need to do some manual stuff
 
So, lets consider an example:
What if we want to allow all users who has `admin` word in email to update posts. Then I'd define rules as this:
```js
can('update', 'Post', { 'author.email': { $regex: "admin" } })
```
In order to retrieve all posts which such user can update we need to create query like this:
```sql
SELECT *
FROM posts as p
  INNER JOIN users as u ON u.id = p.author_id
WHERE u.email LIKE "%admin%"
```
And now your `rulesToQuery` function should be able to detect that `author.email` is a field of another entity and you need to do inner join

**Comment by rconstantine** [comment]
I'm having difficulty understanding this. My scenario: VueJS app with Express backend. For now, I'm looking to use CASL on the frontend at least. I'm using sequelize to connect to MSSQL. I'd like to store rules to the database, and possibly build them in an admin UI.

I think that I should be able to manually build JSON for each rule somehow and store it, then retrieve it and shove it into abilities (multiple abilities can be one each for roles, right?). Is this assumption correct, to start with?

Looking at issue #44, can I just build rules like this:
```javascript
[
  ["read", "Post"],
  ["read", "Book", { "id": 1 }],
  ...
]```

and store them that way?

Or should I use the other way shown:
```javascript
[
  { "actions": "read", "subject": "Post" },
  { "actions": "read", "subject": "Book", "conditions": { id: 1 } },
  ...
]```

Or is this second way not supported anymore? I think it's more readable. What functions do I use to handle these? The packRules/unpackRules shown in issue #44? Or json.stringify/parse? Or toJSON()? Is there a counterpart for toJSON()? If I'm manually making the JSON, I guess I just need the opposite of toJSON?

Is this how I'd extend things?:
```javascript
[
  ["read,edit", "Post"],
  ["read", "Book,Article", { "id": 1 }],
  ...
]```

Or would it be better to extend in this way instead?
```javascript
[
  ["read", "Post"],
  ["edit", "Post"],
  ["read", "Book", { "id": 1 }],
  ["read", "Article", { "id": 1 }],
  ...
]```

And I assume I can extend it for the rest of these:
[
  actions, // string separated by comma
  subjects, // string separated by comma,
  inverted, // true or undefined
  conditions, // object
  fields, // string separated by comma
  reason // string
]

Why is the 'inverted' item skipped in the above examples?

Lastly, I don't know that I would need ruleToQuery in this case, or would I?

**Comment by stalniy** [comment]
@rconstantine #44 was about creating a possibility to minify CASL rules for transporting them through JWT token. Currently there are 2 functions [`packRules/unpackRules`](https://stalniy.github.io/casl/abilities/storage/2017/07/22/storing-abilities.html#abilities-in-jwt-token) which you can use to achieve that but they can't help you with sequelize integration. 

You can read about suggested ways to store rules [here](https://stalniy.github.io/casl/abilities/storage/2017/07/22/storing-abilities.html#storing-abilities). 

There are several ways how to organize rules in database, so you need to decide this by yourself. The simplest usecase is:

1 Role has multiple abilities (one to many relationship). So, you have 2 tables: roles and `abilities`. `abilities` table should have foreign key to `roles` table. Then user can be associated with a role.

Then having a role you can retrieve its rules. 

Alternatively you can use JSON data type for your column in `roles` table and put inside the whole JSON. It should have format like this:

```
[
  { "actions": "read", "subject": "Post" },
  { "actions": "read", "subject": "Book", "conditions": { id: 1 } },
  ...
]
```

**Comment by rconstantine** [comment]
@stalniy Thanks for your last answer. That helped me create the database tables I needed and the administrative screens for those tables. I'm now finally circling back and trying to finish this off. I had to go and do more visible items in my application so management could 'see' progress. Working on hidden stuff like user authorization makes them think you haven't done any work!

In Vue, when I call this.$ability.update(myStuff) where myStuff looks something like:
````javascript
[
  { "actions": "read", "subject": "Post" },
  { "actions": "read", "subject": "Book", "conditions": { id: 1 } },
  ...
]
````

I get an error stating update is not a function. I'm running "@casl/ability": "^2.5.0",  "@casl/vue": "^0.4.3".

I want the above JSON to replace any current rules. Indeed, I also want to be able to empty the rules out using this.$ability.update([]), but that gives the same error. What could I be doing wrong?

**Comment by rconstantine** [comment]
@stalniy, never mind. I figured it out. I had added a hard-coded definition of rules in a js file and initialized the ability object with it. It was those rules I was trying to override. Instead, I just removed them so that the $ability object would begin empty and now the update() function is working without error.

To be clear, I had this in one file:
````javascript
import { abilitiesPlugin } from '@casl/vue'
import { RoleAbilities } from './casl/roleAbilities'

export default ({ Vue }) => {
  Vue.use(abilitiesPlugin, RoleAbilities)
}
````
And this in another:
````javascript
import { AbilityBuilder } from '@casl/ability'

const Administrator = AbilityBuilder.define(can => {
  can('manage', 'all')
})

const Executive = AbilityBuilder.define(can => {
  can('read', 'all')
  can('update', 'Post', { state: 'waitForModeration' })
})

const defaults = AbilityBuilder.define(can => {
  can('read', 'all')
})

export const RoleAbilities = { Administrator, Executive, defaults }
````
I don't know what happened behind the scenes to prevent the update function from working, but changing the first file to no longer use the second seems to have worked. Now I just have to employ the rules.

Thanks again for this module and for the Vue integration.

**Comment by stalniy** [comment]
I don’t plan to create a core package for sequelize as it looks like that project is dying slowly. Instead I plan to add support for [objection.js](https://vincit.github.io/objection.js/) or/and [mikroORM](https://b4nan.github.io/mikro-orm/)

**Comment by musicformellons** [comment]
Mmmh, what makes you think sequelize is dying?
![Screenshot from 2019-06-11 12-20-26](https://user-images.githubusercontent.com/10334123/59264453-6d8a7a00-8c43-11e9-8d7d-3cf1b0b9546c.png)

**Comment by stalniy** [comment]
Probably I used to loud word “dying”). This is just my opinion after working with sequelize for half a year

this is the list of what I don’t like in sequelize:
- no query builder 
- `raw` option just passes generated query to driver, and returns result without proper grouping 
- I had a lot of discussion in my team about using classes for models or not (because documentation didn’t have info about classes but now it has)
- sequelize try to query database with inefficient generated query (e.g., when you have join, limit and offset, it produces query which doesn’t use indexes without any warning or using a better strategy)
- it’s hard to extend sequelize. There is no something like plugins in mongoose, only hooks. 
- in order to insert a graph of objects into db, you need to pass a graph of `include`s

So, for standard use cases it works good, as well as mentioned libraries. But on the other hand objection and mikro-orm allows to work with non standard use cases easier (because both have query builder and more solid codebase)

**Comment by stalniy** [comment]
Sometime ago I [commented in sequelize repo on replacement of string based operators with symbol based](https://github.com/sequelize/sequelize/issues/10820#issuecomment-501959070)

It’s ridiculous, I said that Symbols don't fix the problem, they agreed and closed the issue.

**Comment by ujwal-setlur** [comment]
@stalniy any status on [mikroOrm](https://b4nan.github.io/mikro-orm/) support? I am using mikro-orm and would love this support. Happy to help or test!

**Comment by stalniy** [comment]
No progress for now. But have some thoughts:
1. The best way to start is to integrate with knex
2. MikroORM supports knex, so it should work with knex integration out of the box (or with minor configuration)
3. In knex, we can’t check whether a field is a json field or a foreign key ref (usually we can do this only on ORM level), that’s why we need to introduce extra syntax in conditions that emphasize field as a foreign field ref (to support joins). 
4. To fulfill 3, I’ll need to send a PR in sift. So, this extra syntax in 3 can be processed on casl side (by passing `getter` option, for example)

**Comment by stalniy** [comment]
So, to start we need a function `rulesToKnexQuery(ability, action, subject, options)`. `options` object should contain a property `isRelation: (prop, subjectType, subject) => boolean`. This function  
returns `true` or `false` if `prop` is a relation to another entity. `prop` can be a string (e.g., `name`) or string with dot notation (e.g., `author.email`). If function return `true` we need to `innerJoin` relation and add conditions to `where` clause, if `false` then this is a local field and just add condition to `where` clause.

The result of this `rulesToKnexQuery` is a knex `QueryBuilder` that fetch records in a single request. If we do function like `rulesToKnexQuery` than it will be very easy to add integration for objection.js and mikroORM.js and any knex based ORM.

Afterwards we can even implement something similar for sequelize but in this case the result of the function will be an object, which we can pass to `findAll` because sequelize doesn't have query builder

@ujwal-setlur if you want to try to do this, your help is very welcome!

**Comment by ujwal-setlur** [comment]
@stalniy I will try to get to this in the next week or two!

**Comment by otaviosoares** [comment]
I want to help too. I intend to use it with objection.js. @stalniy do you plan to have `rulesToKnexQuery` at @casl/ability/extra as well?

**Comment by stalniy** [comment]
As a separate package. @casl/knex

**Comment by stalniy** [comment]
I’ll create a folder for it and prepare the infrastructure code

**Comment by otaviosoares** [comment]
Sure, let me know when you want to chat.

**Comment by stalniy** [comment]
> Sure, let me know when you want to chat.

sent you an email with instructions and created infra code for casl-knex in feat/knex branch

**Comment by stalniy** [comment]
Waiting for the feedback from @crcn in https://github.com/crcn/sift.js/issues/202

**Comment by stalniy** [comment]
Guys, is here anybody want to help to integrate casl with sequelize? 

**P.S.:**

I created a new library called [ucast](https://github.com/stalniy/ucast) which can help to do this without efforts.

This library will eventually replace sift.js and will allow to integrate casl with any database system, including ElasticSearch, Casandra and other NoSQL dbs.

**Comment by stalniy** [comment]
Hi @ujwal-setlur ,

You said that can help with mikro-orm integration. Do you still have desire to help?
If so, please check the issue - https://github.com/stalniy/ucast/issues/5

**Comment by ujwal-setlur** [comment]
@stalniy, yes, I can take a stab at it. Thank you for the sequelize gist. I am sure it will be useful. I will take a stab at it this weekend

**Comment by stalniy** [comment]
Hi everybody,

Just some awesome updates regarding work towards SQL support in CASL:

Thanks to outstanding contribution of @otaviosoares,  I've published the first version of [@ucast/objection]() library that helps to translate mongo query into SQL! So, as soon as I [replace siftjs with ucast](https://github.com/stalniy/casl/issues/350), all you will be able to use `objection.js` ORM and CASL to query accessible records. 

It supports auto-joins by the way, so [the case described by me in the comment above](https://github.com/stalniy/casl/issues/8#issuecomment-332150042) is now possible!

Thank you very much @otaviosoares, I really appreciate your help!

**Comment by ShallmentMo** [comment]
@stalniy `as soon as I replace siftjs with ucast` any progress on this?

**Comment by stalniy** [comment]
I'm working on v5 in `next` branch, and plan to publish prerelease version today or tomorrow. So, all feedback is very welcome

**Comment by ShallmentMo** [comment]
looking forward for it 👍

**Comment by stalniy** [comment]
@ShallmentMo release 5.0.0 in next channel. It uses `@ucast/mongo2js` library instead of siftjs. 

The gist how to use objection with CASL right now can be found [here](https://gist.github.com/stalniy/b710d65ac8a6c15f37a435c910624ef7). It's raw however. Now, I need to change `rulesToQuery` (or create `rulesToAST`) function that will produce a final AST which then can be interpreted by `@ucast/objection`. 

But fill free to play with what we have right now. Looking forward to the feedback

**Comment by ShallmentMo** [comment]
Now we are using Sequelize. The gist you write might not fit. So I just write something like this:
```
const { Op } = Sequelize
  const { $and = [], $or = [] } = rulesToQuery(ability, action, object, (rule) => {
    return rule.conditions || {}
  })
  return {
    [Op.and]: $and,
    [Op.or]: $or
  }
```

**Comment by stalniy** [comment]
Guys, we are now very close to have support for SQL! I've just published alpha version of [@ucast/sql](https://www.npmjs.com/package/@ucast/sql) that unifies integration of ucast with 4 major JavaScript ORMS (sequelize, objection, typeorm, mikro-orm). 

Now, ucast conditions can be easily converted to query builder of any of the listed ORMs. That means that casl v5 will support SQL databases!

**Comment by GuskiS** [comment]
Great news. But what about knex?

**Comment by stalniy** [comment]
@GuskiS, in knex we can’t do auto join. So, you can just use @ucast/sql +casl + knex.raw (or knex().whereRaw

**Comment by vemundeldegard** [comment]
@stalniy Can't wait to use this. You said TypeORM is supported, but I can't see it in docs?

**Comment by stalniy** [comment]
> @stalniy Can't wait to use this. You said TypeORM is supported, but I can't see it in docs?

Supported. No worries :) I’ll update readme later. I’m finishing other stuff for v5, you can check the milestone https://github.com/stalniy/casl/milestone/5

casl/sql will be a small wrapper around casl/ability/extra and ucast/sql

So, I think it will take up to month to finish everything and test. I’ll publish prereleases earlier so you will be able to test and provide feedback

**Comment by stalniy** [comment]
Any suggestions regarding API for integration with sequelize and typeorm is very welcome. What I have so far is:

## What I plan to implement

- **@casl/sql/objection** exports: 
  - `AccessibleRecordsQueryBuilder` that has `accessibleBy(ability, action)` method.  
     Integration of `QueryBuilder` can be done by specifying static `QueryBuilder` prop on `Model` (as described [here](https://vincit.github.io/objection.js/recipes/custom-query-builder.html)). 
   
     ```js
     // BaseModel.js
     const { AccessibleRecordsQueryBuilder } = require('@casl/sql/objection');

     class BaseModel extends Model {
       static QueryBuilder = AccessibleRecordsQueryBuilder;
     }

     // Person.js
     const { BaseModel } = require('./BaseModel');

     // Person now uses AccessibleRecordsQueryBuilder
     class Person extends BaseModel {}

     Person.query().accessibleBy(ability).where(...)
     ```
  - `toObjectionQuery(ability, action, objectionQuery)` which adds conditions to specified `objectionQuery` (this function is called by `accessibleBy` internally). This is in case you don't want/can't use custom `AccessibleRecordsQueryBuilder`:
 
     ```js
     const { toObjectionQuery } = require('@casl/sql/objection');

     class Person extends Model {}
  
     const query = toObjectionQuery(abilitly, 'read', Person.query());
     ``` 
- **@casl/sql/mikro-orm** exports:
  - `AccessibleEntityRepository` which can be set as default entity repository for all entities through `MikroORM.init`

    ```js
    const { MikroORM } = require('mikro-orm')
    const { AccessibleEntityRepository } = require('@casl/sql/mikro-orm')

    const orm = await MikroORM.init({
      entities: [UserSchema, ProjectSchema],
      dbName: ':memory:',
      type: 'sqlite',
      entityRepository: AccessibleEntityRepository,
    });

    const query = orm.em.getRepository(User)
      .accessibleBy(ability, action)
      .where(...);
    ```
  - `toMikroORMQuery(ability, action, query)`, similar to objection.js, this function can be used in case you can't/don't want use custom `EntityRepository`

## sequelize

sequelize has [scopes](https://sequelize.org/master/manual/scopes.html) but they look really ugly when you need to pass extra parameters in a scope (`Project.scope('random', { method: ['accessibleBy', ability] })`). So, probably here the best way is just to export `toSequelizeQuery` function and use in Model's static methods when needed:

```js
const { Model } = require('sequelize')
const { toSequelizeQuery } = require('@casl/sql/sequelize')

class Project extends Model {
  static accessibleBy(ability) {
    return this.findAll(toSequelizeQuery(ability, 'read'))
  }
}
```

Any thoughts or feedback is very welcome

## TypeORM

Don't have much knowledge about TypeORM. So, the best I can do is to provide a function `toTypeORMQuery(ability, action, query)`.

```js
const query = toTypeORMQuery(ability, 'read', conn.createQueryBuilder(Person, 'p'))
```

Any thoughts or feedback is very welcome

**Comment by burgalon** [comment]
Are there any commits/starting PR re Sequelize or Sequelize with scopes from your last comment?

**Comment by stalniy** [comment]
not yet. There are several pre-releases for casl v5. It's finished but I plan to create `@casl/sql` based on `@casl/ability` v5 prerelease and `@ucast/sql` prerelease to ensure that they work nice together. 

Also it'd be good to get a feedback about about [this comment](https://github.com/stalniy/casl/issues/8#issuecomment-681191411), just to reduce likelihood of introducing breaking changes on the early stages. I don't use all that ORMs in production, so you guys the best advisors I have currently :)

**P.S.**: the work is also not moving forward because I'm currently looking for a new job and doing test tasks ;) But I think that `@casl/sql` will be available in the beginning of October or earlier.

**Comment by veryeasily** [comment]
I have a few questions regarding the objection.js plugin. The first may be a little bike-shed-y, but would it be possible to provide a class decorator like:

```javascript
// BaseModel.js
const { Accessible } = require('@casl/sql/objection')
const { Model } = require("objection")

class BaseModel extends Accessible(Model) {}

// Person.js
const { BaseModel } = require('./BaseModel')

class Person extends BaseModel {}

Person.query().accessibleBy(ability).where(...)
```

That seems more like it would match up with their [plugin development best-practices](https://vincit.github.io/objection.js/guide/plugins.html#plugin-development-best-practices).

The second question has to do with related queries. My apologies if this one is obvious, but here goes. If I have something like

```javascript
// base_model.js
const { Model } = require("objection");
const { AccessibleRecordsQueryBuilder } = require("@casl/sql/objection");

class BaseModel extends Model {
  static get QueryBuilder() {
    return AccessibleRecordsQueryBuilder;
  }
}

// comment.js
class Comment extends BaseModel {}

// post.js
class Post extends BaseModel {
  static get relationMappings() {
    return {
      comments: {
        relation: Model.HasManyRelation,
        modelClass: Comment,
        from: "Posts.id",
        to: "Comments.postId",
      },
    };
  }
}

// ability.js
const { Ability, AbilityBuilder } = require("@casl/ability");

function defineAbilitiesFor(user) {
  const { can, rules } = new AbilityBuilder();

  can("read", "Post");
  can("read", "Comment", { published: true });

  return new Ability(rules);
}
```

What is the correct way for me to execute something like `Post.query().withGraphFetched("comments")` in such a way where the comments are required to have a true `published` field? Is the idea that `Post.query().accessibleBy(ability).withGraphFetched("comments")` will work out of the box there?

Thanks again for building this! I am super super looking forward to it

**Comment by stalniy** [comment]
Thank you @veryeasily for the thoughtful comment! 

1. Mixin - ok
2. `withGraphFetched` won’t work out of the box. Because the root entity is `Post` and `accessibleBy` gets rules for the post entity. To support `withGraphFetched`, I see 2 options: 
    * using separate method `withAccessibleGraphFetched(ability, 'comments')`
    * using where conditions (not sure about query builder methods): `qb.where(() => Comment.accessibleBy(ability))`

**Comment by veryeasily** [comment]
Thanks for the reply -- FWIW, off the cuff, that 2nd approach with the where conditions looks good to me. It looks hopefully easier to maintain from `casl`'s perspective and also easier for me to reason about from a library user perspective.

**Comment by stalniy** [comment]
casl@5 has been just released. All packages and docs are updated. **SQL support** is coming...

**Comment by lambrohan** [comment]
@stalniy Can I use v5 with postgres and typeorm?

**Comment by stalniy** [comment]
Yes, it’s possible. Also you can manually plug @ucast/sql to get some semi-basic functionality to transform casl rules to sql

**Comment by lambrohan** [comment]
@stalniy Do you have any examples on how can it be done properly?

**Comment by stalniy** [comment]
It’s not battle tested yet. Don’t have enough time to do this. That’s why there is no docs

**Comment by JaneJeon** [comment]
For those of us using Objection.js, please see https://github.com/janeJeon/objection-authorize, I’ve been using and building it for the past few years, _and_ it’s battle tested, _and_ it supports full isomorphism!

**Comment by stalniy** [comment]
Guys, I proud to say CASL is now officially supports [Prisma](https://prisma.io/). So, anyone who uses prisma as ORM can try to use `@casl/prisma` and provide feedback

**Comment by natepelzel** [comment]
Hey, thanks for all the hard work on this project.  I am very interested in the MicroORM integration, and think your idea to handle it via a custom repository makes a lot of sense.

Have you began development on it yet? Is there anything I can do to assist?

**Comment by stalniy** [comment]
Hi @natepelzel 

Use @ucast/sql@alpha version to test integration with mikro-orm. there is a subpackage @ucast/sql/mikro-orm which exposes this functionality.

There are some edge cases in objection.js integration which might be the same for mikro-orm, check this issues:
- https://github.com/stalniy/ucast/issues/17
- https://github.com/stalniy/ucast/issues/18

**Comment by jassdv** [comment]
is there any update regarding Objection support? 
@casl/sql/objection ?
would that be available any time soon?
or something I can start with ?

**Comment by JaneJeon** [comment]
@jassdv unfortunately, the author doesn't seem to see Objection as the right platform to target. In the meantime, however, there is https://github.com/JaneJeon/objection-authorize which has been used for years and provides CASL integration

**Comment by stalniy** [comment]
@JaneJeon this is not true. I’ve never said that. Despite the fact there is no update on the work it doesn’t mean that you can use this thread to promote your library. 

Please or me keep it focused on the actual issues.

**Comment by raymondsze** [comment]
Its good to see if there is official library support typeorm as typeorm itself and nestjs are getting quite popular.

**Comment by kasvith** [comment]
Is there any docs about objection integration?

**Comment by dberardo-com** [comment]
is there any plan to support sequelize as well on top of prisma ? or would you recommend using prisma only with casl while keeping all rest of code with sequelize ? could this actually work this way (mixed approach)?

**Comment by philipimperato** [comment]
> I don’t plan to create a core package for sequelize as it looks like that project is dying slowly. Instead I plan to add support for [objection.js](https://vincit.github.io/objection.js/) or/and [mikroORM](https://b4nan.github.io/mikro-orm/)

This did not age well, haha!

**Comment by dberardo-com** [comment]
and do you know if it is possible to use prisma and sequelize in the same project? basically using prisma only for the policy classes and stuff required by casl while keeping sequelize for all the rest ?

**Comment by MiguelHuamanMedina** [comment]
Hi, I'm trying to implement casl in my node express project with sequelize (using postgres as db), but it's difficult for me to add the conditions part so that the user can only process their records and other cases, has anyone been able to implement it? or what is the best guide to apply the conditions in the casl rules

**Comment by raikusy** [comment]
Did anyone implement CASL with drizzle + postgres? Would appreciate if someone can provide any example or guide.

**Comment by hitesh-pathak** [comment]
CASL v6 docs are not clear on how to use this with say TypeORM, I get that there is no official integration. But from the docs it looks like you can make it work at least for some basic queries. Would like to get any pointers. 

I am using TypeORM & MySQL. Would consider migrating to Prisma but there are some issues with Prisma, like it doesn't provide specialized support for tree entities like TypeORM.

**Comment by josesilveiraa07** [comment]
Any updates for that? I'd love to use it with TypeORM, It's the only blocking issue for me.

**Comment by jd1378** [comment]
maybe [knex](https://www.npmjs.com/package/knex) support ?

**Comment by stalniy** [comment]
Look at @ucast/sql latest alpha version. This is where I stopped and currently I don’t have time to work on new features in open source

**Comment by jd1378** [comment]
Thanks for the work you have done so far, these stuff takes a lot of precious time and I understand that well as I have other open source projects myself, and I wish you the best of luck

It would be great if you could do one more small favor and give me an overview of steps i need to take to make it work with create ability, as it seems ambiguous to me how i can make something like createMongoAbility out of it
If you don't have the time that is fine and all

**Comment by ygrishajev** [comment]
@jd1378 this can be a good starting point: https://gist.github.com/ygrishajev/9ef01444fdb5c386c43b6611400c0fc6
which is based on some findings in this thread above

**Comment by ygrishajev** [comment]
oh, right, confused with possibility of usage with knex

**Comment by jd1378** [comment]
I guess I misunderstood the `createMongoAbility` all along because I had small issues with using it properly
I'm not sure if I understand it correctly now, but after reading the code for it to create one for Sql, I came into conclusion that it is not needed at all, as `createMongoAbility` simply converts a Mongo-like query into a JS runtime check for the ability's can/cannot.

sorry for bothering you all with this notification, just wanted to save someone else's time if they happen to wonder about the same thing coming across this :)

**Comment by araujogui** [comment]
> Did anyone implement CASL with drizzle + postgres? Would appreciate if someone can provide any example or guide.

I created a simple CASL to Drizzle converter. It’s a work in progress and currently missing some features, such as array operators, which I plan to add as soon as possible. Feedback is highly welcome!
Repo: https://github.com/araujogui/ucastle
